wp.hooks - enable the 'all' hook

adamsilverstein opened the pull request:
## Description
Fixes https://github.com/WordPress/gutenberg/issues/8602

## How has this been tested?
First pass: needs testing.

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

adamsilverstein committed:
All action: Shim doAction and applyFilters to trigger the all listener for every hook.

gziolo commented:
What's the status of this one. Is it still something that it expected to be included in the library?

adamsilverstein commented:
> What's the status of this one. Is it still something that it expected to be included in the library?

Yes, this seems genuinely useful (maybe essential) for debugging hooks. 

I'll work on adding some tests and exploring the monkey patching approach forward. To introduce this feature, we need to insure it has minimal or no impact on performance. 

One other option I might explore instead of monkey patching is a debug version of the build that would include the 'all' hook support and would be loaded when SCRIPT_DEBUG is true. that would likely be a cleaner approach and would result in zero impact on production performance. Since this feature is intended for debugging only, this might make the most sense anyway.

What do you think?

gziolo commented:
> One other option I might explore instead of monkey patching is a debug version of the build that would include the 'all' hook support and would be loaded when SCRIPT_DEBUG is true. that would likely be a cleaner approach and would result in zero impact on production performance. Since this feature is intended for debugging only, this might make the most sense anyway.

Sounds great to be disabled by default but enabled in debug mode. I'm wondering how it could be distributed to npm to take advantage of it and ensure that dead code is never bundled for production.

adamsilverstein commented:
> Sounds great to be disabled by default but enabled in debug mode. I'm wondering how it could be distributed to npm to take advantage of it and ensure that dead code is never bundled for production.

How about using the dev/prod build flag so we build the minified production bundle without the 'all' support, while building the unminified version with the 'all' hook support?

I'll try to work on this.

adamsilverstein committed:
add all hook handling wrapped in comment blocks for production removal

adamsilverstein commented:
This small change may be all that is required. I am going to write some tests against this functionality to see if it works as expected. The performance penalty should be very small in practice, when no `all` hooks have been registered a single check is added. The code is wrapped in comment blocks formatted with an eye towards webpack-strip-block - thinking we can strip this off when we create the minified production version of hooks for core.

aduth commented on packages/hooks/src/createRunHook.js:
```diff
@@ -30,6 +30,14 @@ function createRunHook( hooks, returnFirstArg ) {
+		/* develblock:start */
+		// Handle any 'all' hooks registered.
+		if ( hooks.all ) {
+			handlers.push( hooks.all.handlers );
```
You're pushing an array into the array, rather than the members of that array.

i.e.

```
handlers = [ 1 ];
hooks.all.handlers = [ 2 ];

handlers.push( hooks.all.handlers );

console.log( handlers );
// [ 1, [ 2 ] ]
```

You'd want one of either:

```js
handlers = handlers.concat( hooks.all.handlers );
handlers.push( ...hooks.all.handlers );
```

adamsilverstein replied:
eagle eyes! i wouldn't have caught that error until i wrote the tests. will fix

adamsilverstein replied:
this is fixed.

aduth commented on packages/hooks/src/createRunHook.js:
```diff
@@ -30,6 +30,14 @@ function createRunHook( hooks, returnFirstArg ) {
 		const handlers = hooks[ hookName ].handlers;
 
+		// The following code can be stripped from production builds.
+		/* develblock:start */
```
Is this meant to be from a Webpack plugin we're not using?

https://www.npmjs.com/package/webpack-strip-block

We may consider instead a combination of Webpack's (built-in) DefinePlugin and dead-code elimination from its (built-in) minification:

https://webpack.js.org/plugins/define-plugin/

Prior art:

- https://github.com/WordPress/gutenberg/blob/31b6b175af43862dbdcfb9a05ca4887f3ad60116/packages/blocks/src/api/validation.js#L257-L263
- #13324

adamsilverstein replied:
very interesting.

i was actually imagining that the all hook handling would remain in the package by default and that we would _strip it in the WordPress core build for the minified file_. 

In core we already build two versions of each script: one for SCRIPT_DEBUG and one for production. We can use these tags and the webpack-strip-block plugin to remove them in core's production build.  This would completely eliminate the code for produciton and not introduce the overhead of checking the constant.

What do you think?

aduth replied:
The treatment of the environment variable is a bit of a facade, really. It actually achieves the same effect as what you're describing. The way the DefinePlugin works is to literally swap the identifier with some substitute. Thus, if you introduced a condition like this, the body of the condition would be kept for development, and removed for production.

Before (source):

```js
/* a */
if ( process.env.NODE_ENV !== 'production' ) {
	/* b */
}
/* c */
```

After (built with Webpack, unminified):

```js
/* a */
if ( 'production' !== 'production' ) {
	/* b */
}
/* c */
```

After (built with Webpack, minified, comments kept for demonstration only):

```js
/* a *//* c */
```

adamsilverstein replied:
Ok, cool - so we don't need the or the tagging, just a conditional, got it. thanks for clarifying.

So in terms of what we ship in the package, we only ship source right? Do we ship debug scripts with the GB plugin or in core?

my ultimate goal is to have this depend of a wp-config constant, ideally SCRIPT_DEBUG. Do you think its fine keeping this in the pa

aduth replied:
> So in terms of what we ship in the package, we only ship source right? Do we ship debug scripts with the GB plugin or in core?

My expectation would have been that `process.env.NODE_ENV` would be effectively "development" for the non-minified scripts we ship with core, and thus if `SCRIPT_DEBUG` would cause the unminified form to load, it would take effect as you expect. However, considering [an example](https://github.com/WordPress/gutenberg/blob/910b655018037167c2dcee38095905003f632b51/packages/components/src/tooltip/index.js#L94-L97) where we `console.error` a message, I see no reference to said message in the [full unminified source](https://raw.githubusercontent.com/WordPress/WordPress/master/wp-includes/js/dist/components.js) for `wp.components`. I'm not too sure if this is intended or not.

gziolo replied:
I think that `process.env.NODE_ENV` would work the way it's expected. WordPress core would have to generate two versions of the files: `development` and `production`. `SCRIPT_DEBUG` would pick the proper version then. If this isn't the case in core, we should fix it.

adamsilverstein committed:
Add some unit tests for the ‘all’ hook

adamsilverstein committed:
Avoid hook recursion when adding all hook

adamsilverstein commented:
Added some tests demonstrating the 'all' hook in https://github.com/WordPress/gutenberg/pull/12038/commits/92920ffa9b95f7e951df0bda547232e9e6679cf0

adamsilverstein commented:
cc: @johnbillion this might be useful for query monitor

gziolo commented:
@adamsilverstein, do you plan to use guard based on `process.env.NODE_ENV`? I think this is the last thing missing to land this PR.

gziolo reviewed (dismissed):
This PR needs also rebase to trigger new Travis services.

adamsilverstein commented:
> @adamsilverstein, do you plan to use guard based on process.env.NODE_ENV? I think this is the last thing missing to land this PR.

Yes, thanks for the reminder about this ticket. I will work on it.

adamsilverstein committed:
Conditionally add all hook support based on process.env.NODE_ENV

adamsilverstein commented:
@gziolo I rebased and added the switch based on `process.env.NODE_ENV` - does that look right?

adamsilverstein commented:
looks like tests are failing here on the new 'all' hook tests. My guess is this is because the tests run against the production code which does not include the all hook support.

Can we get these tests to run against the development build instead? otherwise I'm not sure how we can test this code?

aduth commented:
If I recall correctly, the tests are run in `NODE_ENV` of test `'test'`. Would it be enough to change the logic from `=== 'development'` to `!== 'production'` ?

gziolo commented:
> If I recall correctly, the tests are run in `NODE_ENV` of test `'test'`. Would it be enough to change the logic from `=== 'development'` to `!== 'production'` ?

100% correct.

adamsilverstein commented:
Great, thanks for the tip - I'll change that.

adamsilverstein committed:
use negative production check

adamsilverstein commented:
@aduth thanks, that fixed it. This is ready for review.

gziolo reviewed (approved):
This all looks good now. Should we include docs for `all` hook and add an entry to CHANGELOG file? It might be worth emphasizing that it works in non-production environments only.

adamsilverstein commented:
> Should we include docs for all hook and add an entry to CHANGELOG file? It might be worth emphasizing that it works in non-production environments only.

Yes, I will work on adding that, thanks!

adamsilverstein committed:
Add a changelog entry describing the all hook

adamsilverstein committed:
Document the `all` hook

adamsilverstein commented:
@gziolo 
* updated the changelog in https://github.com/WordPress/gutenberg/pull/12038/commits/a804f49fd4154bc9208eeead67f98aac191788fc
* updated the readme in https://github.com/WordPress/gutenberg/pull/12038/commits/1a625ec3187b9162cc50f38689db94396e8d6ae0

Can you please review to see if those look good?

gziolo commented on packages/hooks/CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+## 2.1.0 (2019-05-16)
```
```suggestion
## Master
```

gziolo committed:
Update CHANGELOG.md

gziolo commented on packages/hooks/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## Master
```
I updated CHANGELOG myself. We are still not set on the proper recommendation here, but there is new emerging pattern to list new changes as coming from `master`. Probably something good to document. /cc @aduth

gziolo commented:
@adamsilverstein - I slightly modified CHANGELOG (see my comment https://github.com/WordPress/gutenberg/pull/12038#discussion_r284602445). Great work, thanks for wrangling this enhancement 👍
